### PR TITLE
adding promises, starting reorganization of code, adding test

### PR DESCRIPTION
**Pull request by EricH-Propser**
Hey Mark, 

Here are some changes I'm proposing. We need this app to start embracing async a little better. I can go over any of these changes with you and you can merge/deny as you like. 

I want to get to the point where this goes to an mlab store and we can do some fun trophies/achievements with it (fastest answerer, lifetime scores etc)

### CODE REVIEW

File: questions-service.js [hunk 1]
```diff
@@ -8,83 +12,57 @@ var currentAnswersInUse = null;
...
+    for (var question in parsedQuestions) {
+        promises.push(questionValidator.validateQuestion(question)
+                        .catch(function(err){
+                            errorsFound.push("Invalid data for question ["+(i+1)+"] - "+problems.join("; "));
+                        }));
     }
 
+    return Promise.all(promises)
+    .then(function(results){
+        if ( errorsFound.length !== 0 ) {
```
Comment (EricH-Propser) [inline questions-service.js]: If you're not interested in passing back all errors, we can move this to a .catch

**Comment by NeVeX** [comment]
Massive kudo's @EricH-Propser - lemme go through the PR, I don't understand some of it at a glance - but will totally want to adopt these changes - they look sexy!

File: questions-api.js [hunk 2]
```diff
@@ -37,16 +37,17 @@ exports.getAnswerForQuestion = function (request, response) {
...
-           return response.status(201).json({"message": "Successfully updated the questions"})
-       }
-    } else {
+    if ( !request.nevex.isQuizMaster  ) {
         return response.status(403).json({"error": "You are not allowed to set the questions"});
     }
+
+    questionsService.loadQuestionsFromJson(JSON.stringify(request.body))
+    .then(function(result){
+        return response.status(422).json(result);
```
Comment (NeVeX) [inline questions-api.js]: Think this should be the `error` block? Needs to swap with the block below?

File: questions-service.js [hunk 3]
```diff
@@ -154,11 +132,9 @@ exports.generateTestData = function() {
 
     // TODO: make this waaaay better - actually make it more random and support dynamic question answer sizes!
     var totalQuestions = questions.length;
-    var i;
-    for (i = 1; i < 100; i++) {
+    for (var i = 1; i < 100; i++) {
         var name = "test-name-"+i;
-        var q;
-        for ( q = 1; q <= totalQuestions; q++) {
+        for (var q = 1; q <= totalQuestions; q++) {
```
Comment (NeVeX) [inline questions-service.js]: Huh, it didn't let me do that before `for (var q =1 ....)`

File: test/services/question-validator.tests.js [hunk 4]
```diff
@@ -0,0 +1,136 @@
+var validatorDef = require("../../services/question-validator");
```
Comment (NeVeX) [inline test/services/question-validator.tests.js]: Actually amazed that we have tests!

**Pull request merged**